lib/ftp/src/ftp.erl

```diff
@@ -54,6 +54,11 @@
 -export([init/1, handle_call/3, handle_cast/2,
          handle_info/2, terminate/2, code_change/3]).
 
+-export_type([common_reason/0
+             ,restriction_reason/0
+             ,shortage_reason/0
+             ,file_write_error_reason/0]).
+
```

Comment:
I do not like these groups of error atoms, I think it is overkill. `I vote for `atom()` or possibly export one type `ftp:error_codes()`

Reply:
For our users, I agree that maybe `atom()` is enough.
For ourselves, OTP maintainers, I think that being explicit about the possible error type could be useful for Dialyzer to rule out certain possible errors, like mistakes where we could return `{error, foo}`.

I am just not sure yet on how to handle them... One "solution" is that the public interface calls an internal API that has the precise type of error. In this way, the public interface says `{error, atom()}` while the private interface says `{error, common_reason()}`. An example of what I mean follows:

```erlang
-spec user(Pid  :: pid(),
           User :: string(),
           Pass :: string()) ->
    'ok' | {'error', Reason :: atom()}.

user(Pid, User, Pass) -> user_(Pid, User, Pass).
    

-spec user_(Pid  :: pid(),
           User :: string(),
           Pass :: string()) ->
    'ok' | {'error', Reason :: 'euser' | common_reason()}.

user_(_Pid, _User, _Pass) ->
    case {is_name_sane(User), is_name_sane(Pass)} of
        {true, true} ->
            call(Pid, {user, User, Pass}, atom);
        _ ->
            {error, euser}
    end.
```

This allows us to hide the internal return type at the expense of some more lines of information.

Reply:
According to what is said above I think the internal types are useful for the maintainers but the API type I think should be something like possix_atom() | term(). Even if we possibly only return atoms at the moment. I know we have an error section on the man page listing the possix_atoms (or at least the expected ones).